app/models/propertyrepresentation/AppointNewAgentSession.scala

```diff
@@ -16,15 +16,78 @@
 
 package models.propertyrepresentation
 
-import play.api.libs.json.{Json, OFormat}
+import play.api.libs.json.{JsError, JsResult, JsValue, Json, OFormat, Reads, __}
 
-case class AppointNewAgentSession(
+sealed trait AppointNewAgentSession {
+  val status: String
+}
+
+case class SearchedAgent(
+      agentCode: Long,
+      agentOrganisationName: String,
+      agentAddress: String,
+      status: String = "SearchedAgent")
+    extends AppointNewAgentSession
+
+object SearchedAgent {
+  implicit val format: OFormat[SearchedAgent] = Json.format[SearchedAgent]
+}
+
+case class SelectedAgent(
+      agentCode: Long,
+      agentOrganisationName: String,
+      agentAddress: String,
+      isCorrectAgent: Boolean,
+      status: String = "SelectedAgent")
+    extends AppointNewAgentSession
+
+object SelectedAgent {
+  implicit val format: OFormat[SelectedAgent] = Json.format[SelectedAgent]
+
+  def apply(searchedAgent: SearchedAgent, isTheCorrectAgent: Boolean): SelectedAgent =
+    SelectedAgent(
+      agentCode = searchedAgent.agentCode,
+      agentOrganisationName = searchedAgent.agentOrganisationName,
+      agentAddress = searchedAgent.agentAddress,
+      isCorrectAgent = isTheCorrectAgent
+    )
+}
+
+case class ManagingProperty(
       agentCode: Long,
-      agentOrganisationId: Long,
-      agentOrganisationName: Option[String],
-      isCorrectAgent: Option[Boolean],
-      managingProperty: Option[String])
+      agentOrganisationName: String,
+      agentAddress: String,
+      isCorrectAgent: Boolean,
+      managingPropertyChoice: String,
+      status: String = "ManagingProperty")
+    extends AppointNewAgentSession
+
+object ManagingProperty {
+  implicit val format: OFormat[ManagingProperty] = Json.format[ManagingProperty]
+
+  def apply(selectedAgent: SelectedAgent, selection: String): ManagingProperty =
+    ManagingProperty(
+      agentCode = selectedAgent.agentCode,
+      agentOrganisationName = selectedAgent.agentOrganisationName,
+      agentAddress = selectedAgent.agentAddress,
+      isCorrectAgent = selectedAgent.isCorrectAgent,
+      managingPropertyChoice = selection
+    )
+}
 
 object AppointNewAgentSession {
-  implicit val format: OFormat[AppointNewAgentSession] = Json.format[AppointNewAgentSession]
+
+  val readers: Map[String, Reads[_ <: AppointNewAgentSession]] = Map(
+    "SearchedAgent"    -> implicitly[Reads[SearchedAgent]],
+    "SelectedAgent"    -> implicitly[Reads[SelectedAgent]],
+    "ManagingProperty" -> implicitly[Reads[ManagingProperty]]
+  )
+
+  implicit val animalReads: Reads[AppointNewAgentSession] = new Reads[AppointNewAgentSession] {
```

Comment:
animal? also where is the writes.

Reply:
copy and paste error from an example 🤦‍♂. I don't think I need the writes. Will it not use the format from the inheriting case classes ?

Reply:
👍